[WIP] RHDEVDOCS 3910 and RHDEVDOCS-3902- Service binding discoverability and Topology usability improvements (#47760)
openshift/openshift-docs: modules/odc-creating-a-binding-connection-between-components.adoc

:_content-type: PROCEDURE
[id="odc-creating-a-binding-connection-between-components_{context}"]
== Creating a binding connection between components

You can establish a binding connection with Operator-backed components.

> This procedure walks through an example of creating a binding connection between a PostgreSQL Database service and a Spring PetClinic sample application. To create a binding connection with a service that is backed by the PostgreSQL Database Operator, you must first add the Red Hat-provided PostgreSQL Database Operator to the *OperatorHub*, and then install the Operator.
> The PostreSQL Database Operator then creates and manages the `Database` resource, which exposes the binding information in secrets, config maps, status, and spec attributes.

Reviewer: ```suggestion
This procedure walks through an example of creating a binding connection between a PostgreSQL Database service and a Spring PetClinic sample application. 

To create a binding connection with a service that is backed by the PostgreSQL Database Operator, you add the Red Hat-provided PostgreSQL Database Operator to the *OperatorHub*, and then install the Operator. The PostreSQL Database Operator then creates and manages the `Database` resource, which exposes the binding information in secrets, config maps, status, and spec attributes.

When you create the service binding, the {servicebinding-title} controller projects the database service connection information into the application deployment as files by using a volume mount. After the request is successful, the application is redeployed and the connection is established.

.Binding connector
image::odc-binding-connector.png[]
```

Author: It will be covered as part of https://issues.redhat.com/browse/RHDEVDOCS-4234

.Prerequisites

> * You have created and deployed a Spring PetClinic sample application that uses the *Developer* perspective.

Reviewer: ```suggestion
* You have created and deployed a Spring PetClinic sample application by using the *Developer* perspective.
```

Author: Done.

> * You have installed the {servicebinding-title} from the OperatorHub.

Reviewer: ```suggestion
* You have installed {servicebinding-title} from the OperatorHub.
```

Author: Done.

> * You have installed the *Crunchy Postgres for Kubernetes* Operator from the OperatorHub that uses `v5` *Update* channel.

Reviewer: ```suggestion
* You have installed the *Crunchy Postgres for Kubernetes* Operator from the OperatorHub by using the `v5` *Update* channel.
```

> * You have installed the *Crunchy Postgres for Kubernetes* Operator from the OperatorHub by using the `v5` *Update* channel.

Reviewer 2: @smunje1 here you are talking about the crunchy postgres operator and above you are asking the user to install redhat postgresql operator. Use the same operator at both places.

Author: Hello @divyanshiGupta 
Could we discuss your comment as part of Jira: https://issues.redhat.com/browse/RHDEVDOCS-4234.  For this PR, we are focusing currently only on the RHDEVDOCS-3910 and RHDEVDOCS-3902 Jiras.  The "Creating a binding connection between components" section is out of scope for these Jira's. 

If there are no further comments for sections, "Discovering and identifying Operator-backed bindable services" for RHDEVDOCS-3902 and "Verifying the status of your service binding from the Topology view" for RHDEVDOCS-3910, could you please give ack on this PR?

Thank you!

> * You have created and deployed a Crunchy PostgreSQL database instance that uses the *Developer* perspective. This instance has the following four components: `hippo-backup`, `hippo-instance`, `hippo-repo-host`, and `hippo-pgbouncer`.

Reviewer: ```suggestion
* You have created and deployed a Crunchy PostgreSQL database instance by using the *Developer* perspective. This instance has the following components: `hippo-backup`, `hippo-instance`, `hippo-repo-host`, and `hippo-pgbouncer`.
```

.Procedure

> . Switch to the *Developer* perspective and ensure that you are in the appropriate project, for example, `my-petclinic`. In the *Topology* view, hover over the Spring PetClinic sample application to see a dangling arrow on the node.

Author: Done.

> . Click and drag the arrow towards the *hippo* database in the Postgres Cluster to make a binding connection with the Spring PetClinic sample application. A *Create Service Binding* dialog box appears.

Author: Done.

> .. Enter the name and click *Create*.

Reviewer 3: @rolfedh Thanks for that feedback. I created a new issue for our UI: https://bugzilla.redhat.com/show_bug.cgi?id=2107469

> .. Enter the name and click *Create*.

Author: It will be covered as part of https://issues.redhat.com/browse/RHDEVDOCS-4234

+

> .Service Binding modal

Author: Done.

image::odc-sbc-modal.png[]

Alternatively, in the *+Add* view, click the *YAML* option to see the *Import YAML* screen. Use the YAML editor and add the `ServiceBinding` resource:

[source,YAML]
----
apiVersion: binding.operators.coreos.com/v1alpha1
kind: ServiceBinding
metadata:
    name: spring-petclinic-pgcluster
    namespace: my-petclinic
spec:
    services:
    - group: postgres-operator.crunchydata.com
      version: v1beta1
      kind: PostgresCluster
      name: hippo
    application:
      name: spring-petclinic
      group: apps
      version: v1
      resource: deployments

> ----

Author: It will be covered as part of https://issues.redhat.com/browse/RHDEVDOCS-4234

> A service binding request is created and the {servicebinding-title} controller projects the database service connection information into the application deployment as files by using a volume mount. After the request is successful, the application is redeployed and the connection is established.
> 
> .Binding connector
> image::odc-binding-connector.png[]

Reviewer: This is a description of the outcome rather than a step, and doesn't belong inside the procedure. It expands upon the information in line 8. Consider combining this paragraph and illustration with line  8.  
```suggestion
```

Author: It will be covered as part of https://issues.redhat.com/browse/RHDEVDOCS-4234
